src/full/Agda/TypeChecking/Pretty.hs-boot

```diff
@@ -2,20 +2,52 @@
 
 module Agda.TypeChecking.Pretty where
 
--- import Agda.Syntax.Common
+import Data.String (IsString)
+
+import Agda.Syntax.Common (NameId)
 import Agda.Syntax.Internal
 -- import Agda.Syntax.Literal
+import Agda.Syntax.Position (Range)
 
 import Agda.TypeChecking.Monad.Base
+import {-# SOURCE #-} Agda.TypeChecking.Monad.Builtin
+import {-# SOURCE #-} Agda.TypeChecking.Monad.Context (MonadAddContext)
+import Agda.TypeChecking.Monad.Debug (MonadDebug)
+import {-# SOURCE #-} Agda.TypeChecking.Monad.MetaVars (MonadInteractionPoints)
+import {-# SOURCE #-} Agda.TypeChecking.Monad.Signature (HasConstInfo)
+import Agda.Utils.Null (Null)
 import Agda.Utils.Pretty (Doc)
 -- import qualified Agda.Utils.Pretty as P
 
-text                  :: String             -> TCM Doc
-sep, fsep, hsep, vcat :: [TCM Doc]          -> TCM Doc
-($$), (<>), (<+>)     :: TCM Doc -> TCM Doc -> TCM Doc
+text                  :: Monad m => String -> m Doc
+sep, fsep, hsep, vcat :: Monad m => [m Doc] -> m Doc
+($$), (<>), (<+>)     :: Applicative m => m Doc -> m Doc -> m Doc
+
+-- Inlining definitions of MonadReify and MonadAbsToCon to avoid
+-- having to import them
+type MonadPretty m =
+  ( ( MonadReduce m
+    , MonadAddContext m
+    , MonadInteractionPoints m
+    , MonadFresh NameId m
+    , HasConstInfo m
+    , HasOptions m
+    , HasBuiltins m
+    , MonadDebug m
+    )
+  , ( MonadTCEnv m
+    , ReadTCState m
+    , MonadStConcreteNames m
+    , HasOptions m
```

Comment:
Interesting, you can nest tuples.
There are duplicates, though.

Reply:
Well ideally I'd just write the actual definition of `MonadPretty` here, which is
```
type MonadPretty m =
  ( MonadReify m
  , MonadAbsToCon m
  , IsString (m Doc)
  , Null (m Doc)
  )
```
However this requires some extra imports that cause import loops in the boot file. So I just iota-expanded the definitions of `MonadReify` and `MonadAbsToCon`. It's not pretty, but I don't know a better solution...